docs/howto/how-to-make-tensorflow-plugin-with-ui/tensorflow-plugin-with-ui.md

```diff
@@ -0,0 +1,1488 @@
+
+\* Note the distinction here as well. A _log_ statement will automatically send your output to `PythonSdk.log`. A _message_ (read: `self.provider.io.[info | warn | error]`) will send your data output to Designer's IO.
+
+If you have multiple distinct sorts of errors or want to use your Exception derivatives, we recommend defining [decorators](https://pythonguide.readthedocs.io/en/latest/python/decorator.html) to make your exception handling **short, simple, and reusable across any tool you develop as a python module!**
+
+#### Adding a custom method: `setup_data()`
+
+Now we will add our first custom method to our plugin, `setup_data(self)`.
+It may be tempting to write your `DATA` mode code exactly where it will execute.
+**However** we strongly encourage following general coding best practices.
+Think, "How would I do this **in Python**?"
+Rather than, "How do I do this in the `ayx_python_sdk`?"
+When developing with the SDK, we encourage developers to use Python to its fullest potential.
+Therefore, we place our `DATA` mode code in its function, `setup_data(self)`.
+Separating your steps into functions naturally helps isolate problem areas in your code, makes debugging much easier, and (like mentioned above) makes even **more** of your code reusable! 
+
+```python
+  def setup_data(self):
+    train_dir = self.provider.tool_config["datasetConfig"]["trainingSetDir"]
+    test_dir = self.provider.tool_config["datasetConfig"]["testSetDir"]
+
+    seed = int(self.provider.tool_config["datasetConfig"]["seed"])
+    batch_size = int(self.provider.tool_config["datasetConfig"]["batchSize"])
+
+    if Path(train_dir).exists():
+      raw_train_ds = tf.keras.utils.text_dataset_from_directory(
+        train_dir, 
+        batch_size=batch_size, 
+        validation_split=0.2, 
+        subset='training', 
+        seed=seed)
+      raw_test_ds = tf.keras.utils.text_dataset_from_directory(
+        test_dir,
+        batch_size=batch_size)
+      raw_val_ds = tf.keras.utils.text_dataset_from_directory(
+        train_dir,
+        batch_size=batch_size, 
+        validation_split=0.2, 
+        subset='validation', 
+        seed=seed)
+      
+      try:
+        tf.data.Dataset.save(raw_train_ds, f"{self.data_url}/train")
+        tf.data.Dataset.save(raw_test_ds, f"{self.data_url}/test")
+        tf.data.Dataset.save(raw_val_ds, f"{self.data_url}/validation")
+        self.info("Completed data prep...")
+      except Exception as e:
+        logger.error(repr(e))
+```
+We know already `self.provider.tool_config` is simply a Python dict.
+Note, though, we have to call `int` on some values. We do this simply because we get all `tool_config` as `string` types from the UI SDK.
+So, we need to convert this to their expected types or even serialize more complex data. (that is enough for its step-by-step guide)
```

Comment:
TW review:

We know `self.provider.tool_config` is a Python dict. Note, though, we have to call `int` on some values. We do this because we get all `tool_config` as `string` types from the UI SDK. So, we need to convert this to their expected types or even serialize more complex data.